src/Raven.Server/Extensions/CancellationTokenSourceExtensions.cs

```diff
@@ -0,0 +1,121 @@
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace Raven.Server.Extensions
+{
+    public static class CancellationTokenSourceExtensions
+    {
+        public static CancellationTimeout TimeoutAfter(this CancellationTokenSource cts, TimeSpan dueTime)
+        {
+            return new CancellationTimeout(cts, dueTime);
+        }
+    }
+
+    public class CancellationTimeout : IDisposable
+    {
+        public CancellationTokenSource CancellationTokenSource { get; private set; }
+        private readonly Timer _timer;
+        private readonly int _dueTime;
+        private readonly object _locker = new object();
+        private volatile bool _isTimerDisposed;
+        private readonly Stopwatch _sp = Stopwatch.StartNew();
+
+        public void ThrowIfCancellationRequested()
+        {
+            CancellationTokenSource.Token.ThrowIfCancellationRequested();
+        }
+
+        public CancellationTimeout(CancellationTokenSource cancellationTokenSource, TimeSpan dueTime)
+        {
+            if (cancellationTokenSource == null)
+                throw new ArgumentNullException(nameof(cancellationTokenSource));
+            if (dueTime < TimeSpan.Zero)
+                throw new ArgumentOutOfRangeException(nameof(dueTime));
+
+            _isTimerDisposed = false;
+            CancellationTokenSource = cancellationTokenSource;
+            _dueTime = (int) dueTime.TotalMilliseconds;
+            _timer = new Timer(self =>
```

Comment:
In the querying we could use `CancellationTokenSource` with an appropriate timeout directly:
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/CancellationTokenSource.cs/#L273
or `CancelAfter` method and this way do not create `CancellationTimeouts` at all (we don't need `Delay/Pause/Resume` in that scenario anyway).

However, under the hood they also use timers. The creation  (and disposal) of 1,000,000 such cancellation tokens takes 00:00:00.3251197. It doesn't appear to be a significant cost per query.